File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/properties/PropertiesSpec.java
package io.micronaut.rabbitmq.docs.properties;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;

@MicronautTest
@Property(name = "spec.name", value = "PropertiesSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class PropertiesSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testPublishingAndReceivingProperties(ProductClient productClient, ProductListener productListener) {
// tag::producer[]
        productClient.send("body".getBytes());
        productClient.send("guest", "text/html", "body2".getBytes());
        productClient.send("guest", null, "body3".getBytes());
// end::producer[]


        await().atMost(60, SECONDS).until(() ->
                productListener.messageProperties.size() == 3 &&
                productListener.messageProperties.contains("guest|application/json|myApp") &&
                productListener.messageProperties.contains("guest|text/html|myApp") &&
                productListener.messageProperties.contains("guest|null|myApp")
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/properties/PropertiesSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="PropertiesSpec")
class PropertiesSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_publishing_and_receiving_properties(self) -> None:
# tag::producer[]
        self.product_client.send(b"body")
        self.product_client.send_with_properties("guest", "text/html", b"body2")
        self.product_client.send_with_properties("guest", None, b"body3")
# end::producer[]

        message_properties = self.product_listener.message_properties
        Await.until(lambda: len(message_properties) == 3
                    and "guest|application/json|myApp" in message_properties
                    and "guest|text/html|myApp" in message_properties
                    and "guest|None|myApp" in message_properties)
